### PR TITLE
Task 52

### PR DESCRIPTION
**Pull request by nomeata**
Very first steps. Some things were easy generalization of code that was already there. Left some comments about things I'll have to think about.

### CODE REVIEW

File: Carleson/DiscreteCarleson.lean [hunk 1]
```diff
@@ -9,6 +10,127 @@ open Classical -- We use quite some `Finset.filter`
...
+  refine mem_iUnion₂_of_mem hl hp
+
+lemma PairwiseDisjoint_foo : univ.PairwiseDisjoint (foo s) := fun _ _ _ _ => Disjoint_foo
+
+lemma foo_subset {n : ℕ} : foo s n ⊆ s := calc
+  foo s n = minimals (·≤·) (s \ ⋃ (n' < n), foo s n') := foo.eq_def ..
+  _ ⊆ s \ ⋃ (n' < n), foo s n' := minimals_subset ..
+  _ ⊆ s := diff_subset
+
+-- How do we want to express “has no chains of length n+1”?
```
Comment (fpvandoorn) [inline Carleson/DiscreteCarleson.lean:35]: `\all t \sub s, IsChain (\. \le \.) t \to t.ncard \le n`? Not sure about `Set.ncard`.

File: Carleson/DiscreteCarleson.lean [hunk 2]
```diff
@@ -9,6 +10,127 @@ open Classical -- We use quite some `Finset.filter`
...
 
 
+section PartialOrderConstructionWithoutNameSoFar
+
+variable {α} [PartialOrder α] {s : Set α}
+
+-- This operation needs a name.
+-- And maybe easier to define via the height, rather than iteratively?
+
+def foo (s : Set α) (n : ℕ) : Set α := minimals (·≤·) (s \ ⋃ (n' < n), foo s n')
```
Comment (fpvandoorn) [inline Carleson/DiscreteCarleson.lean:20]: maybe `iteratedMinimals`?

**Review by fpvandoorn — COMMENTED** [review]
Also, please move the generic partial-order stuff to a separate file in `ToMathlib`.

File: Carleson/DiscreteCarleson.lean (hunk 1 above)
Comment (fpvandoorn) [reply Carleson/DiscreteCarleson.lean:35]: Oh, I didn't know about `KrullDimension`. That is probably better to use.

File: lakefile.lean [hunk 3]
```diff
@@ -32,6 +32,9 @@ require mathlib from git
 require checkdecls from git
   "https://github.com/PatrickMassot/checkdecls.git" @ "master"
 
+require loogle from git
+  "https://github.com/nomeata/loogle.git" @ "master"
+
```
Comment (nomeata) [inline lakefile.lean:37]: Of course I’ll remove this from this branch when and if this PR is getting closer to being merged.

File: Carleson/ToMathlib/Height.lean [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+      simpa only [mem_minimals_iff_forall_lt_not_mem'', Set.mem_setOf_eq, hfin, le_top, not_le,
+        true_and, ENat.top_ne_coe, iff_false, not_forall, Classical.not_imp, not_lt]
+    simp only [not_lt, top_le_iff, height_eq_top_iff] at hfin
+    obtain ⟨p, rfl, hp⟩ := hfin (n+1)
+    use p.eraseLast.last, RelSeries.eraseLast_last_rel_last _ (by omega)
+    simpa [hp] using height_last_ge_length p.eraseLast
+
+-- Q: Should this be the definition and the other a lemma? Does it matter?
+-- Q: What's a good name?
+def Set.with_height (s : Set α) (n : ℕ) : Set α :=
```
Comment (fpvandoorn) [inline Carleson/ToMathlib/Height.lean:414]: `Set.withHeight` is a good name IMO (we use `camelCase` for definitions)

File: Carleson/ToMathlib/Height.lean [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+  ext y
+  wlog hys : y ∈ s
+  · simp [hys]
+  simp only [mem_diff, hys, mem_iUnion, exists_prop, not_exists, not_and, true_and, mem_setOf_eq,
+    exists_and_left, exists_true_left]
+  cases height (⟨y, hys⟩:s)
+  next => simp
+  next z =>
+    simp only [Nat.cast_inj, Nat.cast_le]
+    -- no single tactic for goal `∀ x < n, ¬z = x) ↔ n ≤ z`?
```
Comment (fpvandoorn) [inline Carleson/ToMathlib/Height.lean:464]: [forall_lt_iff_le](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Basic.html#forall_lt_iff_le) is close, but not quite there...

File: Carleson/ToMathlib/Height.lean [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+lemma Set.PairwiseDisjointSet.with_heig_with_height (s : Set α) : univ.PairwiseDisjoint s.with_height :=
+    fun _ _ _ _ => Disjoint_with_height s
+
+/-
+If all increasing series have lenght bounded by `n`, then `s` is the union of its elements with
+height `≤ n`.
+
+The precondition could also be expressed as `(hkrull : krullDim α < n)`.
+-/
+lemma Set.iUnion_with_height_of_bounded_series {s : Set α} {n : ℕ}
```
Comment (fpvandoorn) [inline Carleson/ToMathlib/Height.lean:491]: Is this an `Iff`?

Maybe also state that if all chains are smaller than `n` then `s.withHeight n = \empty`?

**Review by fpvandoorn — COMMENTED** [review]
I'm happy to merge this quickly, and you can open a PR on top of it to clean/finish things.

**Comment by nomeata** [comment]
Ok, for a quick-merge-cleanup-later this might be just about viable now.

**Comment by nomeata** [comment]
Thanks for the cleanup, I was a bit in a rush, it seems. Also forgot to change `with_coheight` to `withCoheight` sorry for that.

**Comment by fpvandoorn** [comment]
No worries. The standards here are much lower than Mathlib, so when you have time, please make a follow-up PR fixing it :-)

**Pull request merged**